src/go/cmd/review.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"slices"
+	"sort"
+	"strings"
+)
+
+var errExternalModule = errors.New("reviewed module exports a type defined in a different repository")
+
+// Review represents an apiview review of an Azure SDK for Go module
+type Review struct {
+	// modules maps module paths to Modules implicated in this API review. It
+	// contains only the reviewed module in most cases, however it contains
+	// more when the reviewed module exports types defined in another module.
+	modules map[string]*Module
+	// name of the APIView review e.g. "sdk/azcore"
+	name string
+	// path on disk to the reviewed module
+	path string
+	// reviewed is the module being reviewed
+	reviewed *Module
+}
+
+// NewReview creates a Review for the module at path p
+func NewReview(p string) (*Review, error) {
+	m, err := NewModule(p)
+	if err != nil {
+		return nil, err
+	}
+	r := &Review{
+		modules: map[string]*Module{},
+		name:    getPackageNameFromModPath(m.ModFile.Module.Mod.Path),
+		path:    p,
+	}
+	fmt.Println("Reviewing", r.name)
+	err = r.AddModule(m)
+	return r, err
+}
+
+// AddModule adds a module to the review. Call this to add a module that exports
+// a type the reviewed module exports by alias.
+func (r *Review) AddModule(m *Module) error {
+	if _, ok := r.modules[m.ModFile.Module.Mod.Path]; ok {
+		return fmt.Errorf("module %s already exists in index", m.ModFile.Module.Mod.Path)
+	}
+	if len(r.modules) == 0 {
+		r.reviewed = m
+	}
+	r.modules[m.ModFile.Module.Mod.Path] = m
+	return nil
+}
+
+func (r *Review) Review() (PackageReview, error) {
+	r.resolveAliases()
+
+	tokenList := &[]Token{}
+	nav := []Navigation{}
+	diagnostics := []Diagnostic{}
+	packageNames := []string{}
+	for name, p := range r.reviewed.Packages {
+		// we use a prefixed path separator so that we can handle the "internal" module.
+		//  internal/dig
+		//  internal/errorinfo
+		//  etc.
+		// for other modules, we skip /internal subdirectories
+		//  azcore/internal/...
+		if strings.Contains(p.relName, "/internal") || p.c.isEmpty() {
```

Comment:
Do we pop as error if a module is exporting a type from its internal package. I would consider it a bug if azure exported a type from azcore/internal

Reply:
We make heavy use of this pattern to avoid circular references.

Reply:
Ah ok, do we enforce that in the APIView tool. We should. Also we should add that to the GO design guidelines.

Reply:
We surface it, but don't enforce it in APIView, and I don't think that it's the correct tool to do so (AFAIK there's no way to fail the APIView).

Reply:
APIView can show errors/warnings, we do it in the C++ APIView.

Reply:
Interesting. How does that impact the APIView, i.e. is it just some markup or does it prevent the APIView from being approved?

Reply:
Its just markup. APIViews simplify show diffs, notes, comments, ... At the end of the day the architects can click Approve whenever they want

Reply:
We do the same for every type alias in Go, and the messages are just as illegible in our reviews 🙈

Reply:
Perfect we can leverage the pattern and the code. Have it show a message that the `type` from internal is being exported. 
   "Type `x` is exported from internal. Exported types from internal must be in the ...../internal/exported module"

Reply:
I opened #8696. We don't consistently use this pattern today, so step 1 of the issue is determining whether we can.